app/services/rate_limiter.rb

```diff
@@ -69,18 +87,37 @@ def maxed?
     attempts && attempts >= RateLimiter.max_attempts(rate_limit_type)
   end
 
+  def expiration_minutes
+    minutes = RateLimiter.attempt_window_in_minutes(rate_limit_type)
+    exponential_factor = RateLimiter.attempt_window_exponential_factor(rate_limit_type)
+    attempt_window_max = RateLimiter.attempt_window_max_in_minutes(rate_limit_type)
+    minutes *= exponential_factor ** (attempts - 1) if exponential_factor && attempts.positive?
+    if attempt_window_max && minutes > attempt_window_max
+      attempt_window_max
+    else
+      minutes
+    end
+  end
+
   def increment!
     return if limited?
     value = nil
 
+    expiration_seconds = RateLimiter.attempt_window_in_minutes(rate_limit_type).minutes.in_seconds
+    exponential_factor = RateLimiter.attempt_window_exponential_factor(rate_limit_type)
+    attempt_window_max = RateLimiter.attempt_window_max_in_minutes(rate_limit_type)
     now = Time.zone.now
     REDIS_THROTTLE_POOL.with do |client|
-      value, _success = client.multi do |multi|
-        multi.incr(key)
-        multi.expireat(
-          key,
-          now + RateLimiter.attempt_window_in_minutes(rate_limit_type).minutes.in_seconds,
+      begin
+        value = client.evalsha(
+          INCREMENT_SCRIPT_SHA1,
+          [key],
+          [now.to_i, expiration_seconds, exponential_factor, attempt_window_max].map(&:to_s),
         )
+      rescue Redis::CommandError => error
+        raise error unless error.message.start_with?('NOSCRIPT')
+        client.script(:load, INCREMENT_SCRIPT)
+        retry
```

Comment:
```suggestion
        client.eval(
          INCREMENT_SCRIPT,
          [key],
          [now.to_i, expiration_seconds, exponential_factor, attempt_window_max].map(&:to_s),
        )
```

My understanding is we can execute the script directly and it'll be put in the cache ([docs](https://redis.io/docs/latest/develop/interact/programmability/eval-intro/#script-cache))

Reply:
Updated in 7dc130ce36